### ISSUE TITLE
Footprint not cleared properly

### ISSUE THREAD

**Issue by reinzor** [SteveMacenski/nonpersistent_voxel_layer#4]
Footprint clearing is working with the ObstacleLayer but not working with this plugin.

**Comment by SteveMacenski** [comment]
Keep in mind what this layer is actually doing - its only marking the most recent sets of measurements at each update iteration, then clearing out the voxel grid again for the next cycle. It wouldn't make much sense to persist information about the footprint to clear. 

If there's a use-case I'm missing, please let me know (and a PR is welcome).

**Comment by reinzor** [comment]
Hi @SteveMacenski , thanks for your reply. I am aware of what this layer does. However, in the default obstacle layer, the footprint is always cleared from the costmap, this does not work properly with this plugin. Our scanner measures parts of our robot sometimes so I would like to exclude the footprint from the costmap.

**Comment by SteveMacenski** [comment]
Ah ok, that makes sense. Let me think about that for a bit. If you have any potential causes, I'd love to hear. 

Lets get that sorted out.

**Comment by SteveMacenski** [comment]
Can you send me your config file? 

I'm looking at obstacle layer's update here https://github.com/ros-planning/navigation/blob/kinetic-devel/costmap_2d/plugins/obstacle_layer.cpp#L415 and it checks for `footprint_clearing_enabled_` which I dont specifically deal with, that's upstream configs. Then I call `updateFootprint`  and `updateCosts` from the obstacle layer. All this is contingent on configs for that being correct

**Comment by reinzor** [comment]
```
footprint:
  - [-0.4, -0.33]
  - [-0.4, 0.33]
  - [0.4, 0.33]
  - [0.4, -0.33]
footprint_padding: 0.05

obstacles_layer:
  mark_threshold: 0
  combination_method: 0
  origin_z: 0
  z_resolution: 0.05
  z_voxels: 16
  publish_voxel_map: true
  observation_sources: front_scan rear_scan
  front_scan:
    data_type: LaserScan
    topic: front_scan_filtered
    min_obstacle_height: -2.0
    max_obstacle_height: 2.0
  rear_scan:
    data_type: LaserScan
    topic: rear_scan_filtered
    min_obstacle_height: -2.0
    max_obstacle_height: 2.0

inflater_layer:
  inflation_radius: 1

rolling_window: true

update_frequency: 20.0
publish_frequency: 10.0

width: 5.0
height: 5.0
resolution: 0.025

plugins:
  - {name: obstacles_layer, type: "costmap_2d::NonPersistentVoxelLayer"}
  - {name: inflater_layer, type: "costmap_2d::InflationLayer"}
```

I already looked into this and I indeed saw that the behavior should be the same, however, it is behaving differently than the default obstacle layer. In the config above, by changing the `NonPersistentVoxelLayer` to `ObstacleLayer`, you will notice the difference.

**Comment by SteveMacenski** [comment]
Hi,
Just pinging back, I recognize this is still an issue and I'm hoping to dive into it during the Christmas/New Years holidays. Just wanting to let you know I haven't forgot. It's just been lower priority to debug as we dont have a configuration that would make use of it.

**Comment by reinzor** [comment]
Thanks for the update :)

**Comment by SteveMacenski** [comment]
In the mean time if you had a relatively small bag file that has data I could use to debug that would be appreciated since I dont have a robot who marks in the footprint.

**Comment by SteveMacenski** [comment]
@reinzor I've looked into this quite a bit

I dont see the same issue you describe, I set up my costmap to look at only this layer and I see this 
![image](https://user-images.githubusercontent.com/14944147/50615799-d0a90100-0e9a-11e9-93ad-841ea2e9f572.png)

That clearly shows the footprint of my robot being cleared out in free space where the rest of it is unknown

**Comment by SteveMacenski** [comment]
I do notice however that when I use your footprint parameters, it clears the first iteration then never again... interesting. Might be something to do with the parsing built in for circular robots vs. polygon specified robots. 

I found a work around, PR to follow. I'll tag you in it, it would be good if you tested and said "ok" then I'll merge and we'll get that shipped out for you.

**Comment by SteveMacenski** [comment]
@reinzor 

Here's the PR https://github.com/SteveMacenski/nonpersistent_voxel_layer/pull/5

Once you give me the OK I'll release another version of it to `apt`

**Comment by reinzor** [comment]
Thanks for getting back to this. I can verify that this works:

`master`:
![old](https://user-images.githubusercontent.com/4176023/50648272-0d4c2980-0f7b-11e9-943a-5cbfd4c0560f.png)

`footprint`:
![new](https://user-images.githubusercontent.com/4176023/50648282-150bce00-0f7b-11e9-8d09-f694f9bd6935.png)

Thanks for the PR!

**Comment by SteveMacenski** [comment]
Thanks for the ticket!

**Comment by SteveMacenski** [comment]
@reinzor 

Will be built in shadow fixed this afternoon https://github.com/ros/rosdistro/pull/19856

Should be in the next kinetic sync
